### PR TITLE
Update IntegralArrays.jl

### PR DESCRIPTION
**Pull request by jakewilliami**
It looks like the code for the integral array is outdated (actually, removed).  Here's a possibility for this code, which I wrote for my face detection algorithm.  I suspect there are things that can be made much better, but this is a start.

### CODE REVIEW

File: src/IntegralArrays.jl [hunk 1]
```diff
@@ -1,5 +1,74 @@
...
+#=
+Any one pixel in a given image has the value that is the sum of all of the pixels above it and to the left.
+Original    Integral
++--------   +------------
+| 1 2 3 .   | 1  3  6 .
+| 4 5 6 .   | 5 12 21 .
+| . . . .   | . . . . .
+=#
+
+function getImageMatrix(imageFile::AbstractString)
```
Comment (timholy) [inline src/IntegralArrays.jl:18]: We follow the [Julia style guide](https://docs.julialang.org/en/latest/manual/style-guide/#Use-naming-conventions-consistent-with-Julia-base/-1) and will probably transition to the [Blue style guide](https://github.com/invenia/BlueStyle) as well, so please using naming consistent with standard Julia choices.

File: src/IntegralArrays.jl [hunk 2]
```diff
@@ -1,5 +1,74 @@
...
+    img = load(imageFile)
```
Comment (timholy) [inline src/IntegralArrays.jl:19]: I am not sure we need IO functionality in this library. There's nothing here specific to an integral array representation.

File: src/IntegralArrays.jl [hunk 3]
```diff
@@ -1,5 +1,74 @@
...
+
+    # imgArr = convert(Array{Float64}, channelview(img)) # for coloured images
+    imgArr = convert(Array{Float64}, Colors.Gray.(img))
+    
+    return imgArr
+end
+
+function toIntegralImage(imgArr::AbstractArray)
+    
+    arrRows, arrCols = size(imgArr) # get size only once in case
```
Comment (timholy) [inline src/IntegralArrays.jl:30]: Note this assumes 2d. While not fatal, most of the algorithms in JuliaImages support 3d images too. Julia makes it easy to use the same code for arbitrary dimensionality.

File: src/IntegralArrays.jl [hunk 4]
```diff
@@ -1,5 +1,74 @@
...
+    rowSum = zeros(arrRows, arrCols)
+    integralImageArr = zeros(arrRows, arrCols)
```
Comment (timholy) [inline src/IntegralArrays.jl:32]: This assumes grayscale but the input arguments are general.

File: src/IntegralArrays.jl [hunk 5]
```diff
@@ -1,5 +1,74 @@
...
+    
+    # process the first column
+    for x in 1:(arrRows)
```
Comment (timholy) [inline src/IntegralArrays.jl:35]:
```suggestion
    for x in 1:arrRows
```

However, to support arbitrary axes (https://julialang.org/blog/2017/04/offset-arrays/) it's best to say `for x in axes(imgArr, 1)`

File: src/IntegralArrays.jl [hunk 6]
```diff
@@ -1,5 +1,74 @@
...
+            integralImageArr[x, 1] = integralImageArr[x-1, 1] + imgArr[x, 1]
+        end
+    end
+    
+    # start processing columns
+    for y in 1:(arrCols)
+        # same as above: we cannot access a 0th element in the matrix
+        # our scalar accumulator s will catch the 1st row, so we only
+        # needed to predefine the first column before this loop
+        if isone(y)
```
Comment (timholy) [inline src/IntegralArrays.jl:50]: Here it would be easier to use `for y in 2:arrCols`, or for more general axes `for y in Iterators.drop(axes(y, 2), 1)`.

**Comment by timholy** [comment]
Actually, see #4. That's a pretty good implementation, so it would probably be best just to modernize that and fix the remaining unaddressed review comments. If you tackle that, make sure you preserve @mronian's contributions under his name.

**Comment by jakewilliami** [comment]
I updated #1 to work for 1.5.  What do you think?  You undoubtedly know much more Julia than I, so please let me know what still needs to be changed.  Including (and especially) regarding code style and convention :)

**Comment by timholy** [comment]
Hi @jakewilliami, sorry for the long delay here. In general when you're updating someone else's PR, it's best to start from their branch so you can preserve the attribution of their contribution.

I merged #1 and credited you as a co-author of my changes. Thanks for helping get this over the finish line!

**Pull request closed without merging**